torchmultimodal/models/gpt.py

```diff
@@ -28,12 +29,278 @@ class TransformerLayerOutput(NamedTuple):
+        attn_mask (Tensor, optional): Tensor of dimension ``(b, q_seq_len, k_seq_len)`` where prefixes ``q``
+            and ``k`` stand for query and key. Contains 1s for positions to attend to and 0s for masked positions.
+            Defaults to ``None``.
+        head_mask (Tensor, optional): Tensor of dimension ``(b, h, q_seq_len, k_seq_len)``. Masks need to be
+            specified for each attention head. Defaults to ``None``.
+        use_cache (bool, optional): If ``True``, caches past key/value tensors for faster decoding. If ``False``,
+            recomputes key and value for each decoding step. Defaults to ``False``.
+        causal (bool. optional): If ``True``, use causal attention. Defaults to ``False``.
+        right_shift (bool): If ``True``, shifts the embedding vectors to the right and prepends it with start of
+            sentence token. Defaults to ``False``. This option is disregarded during training mode
+        return_attn_weights (bool, optional): If ``True``, returns attention probabilities of each transformer
+            layer. Defaults to ``False``.
+        return_hidden_states (bool): If ``True``, returns the embeddings of each transformer layer. Defaults to
+            ``False``.
+
+    Raises:
+        AttributeError: If input tokenizer does not implement methods ``encode`` and ``lookup`` or if output
+            tokenizer does not implement methods ``encode``, ``lookup`` and ``decode``.
+    """
+
+    def __init__(
+        self,
+        d_model: int,
+        num_tokens: int,
+        latent_shape: Tuple[int, ...],
+        in_tokenizer: nn.Module,
+        out_tokenizer: nn.Module,
+        mm_decoder: nn.Module,
+        in_projection: Optional[nn.Module] = None,
+        out_projection: Optional[nn.Module] = None,
+        norm_layer: Optional[Callable[..., nn.Module]] = None,
+    ) -> None:
+        super().__init__()
+        if not all(
+            [hasattr(in_tokenizer, attr_name) for attr_name in ["encode", "lookup"]]
+        ):
+            raise AttributeError(
+                "Input modality tokenizer must have methods 'encode' and 'lookup'."
+            )
+
+        if not all(
+            [
+                hasattr(out_tokenizer, attr_name)
+                for attr_name in ["encode", "lookup", "decode"]
+            ]
+        ):
+            raise AttributeError(
+                "Output modality tokenizer must have methods 'encode', 'lookup' and 'decode'."
+            )
+
+        self.latent_shape = latent_shape
+        self.in_tokenizer = in_tokenizer
+        self.out_tokenizer = out_tokenizer
+        self.mm_decoder = mm_decoder
+        self.in_projection = in_projection
+        self.out_projection = out_projection
+        if norm_layer is None:
+            norm_layer = partial(nn.LayerNorm, eps=1e-5)
+        self.norm = norm_layer(normalized_shape=d_model)
+        self.to_logit = nn.Linear(d_model, num_tokens, bias=False)
```

Comment:
nit: could just call this logit_projection or something

Reply:
I will keep it this way for now as it will make my life easier when re-mapping MUGEN's checkpoints. :)